### PR TITLE
ATO-753: Set up pipeline to deploy to GitHub container registry

### PR DESCRIPTION
**Pull request by gn-GDS**
## What

- Auto tagging on push to main (according to [discovery](https://docs.google.com/document/d/1u-xKH3b19CdA4XTpeKAGM53clWtdHL0kJYmN-uE3dKc/edit))

## How to review
- Check that package release and tags follow agreed formatting

## Checklist

- [X] Impact on orch and auth mutual dependencies has been checked.

### CODE REVIEW

File: .github/workflows/publish-image.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+on:
+  pull_request:
+    types:
+      - opened
+      - reopened
+      - ready_for_review
+      - synchronize
+  merge_group:
+  workflow_dispatch:
```
Comment (CarlyG55) [inline .github/workflows/publish-image.yml:11]: I think this should be on push to main?

File: .github/workflows/publish-image.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+env:
+  REGISTRY: ghcr.io
+  IMAGE_NAME: ${{ github.repository }}
+
+jobs:
+  push_to_registry:
+    name: Push Docker image to Docker Hub
```
Comment (CarlyG55) [inline .github/workflows/publish-image.yml:19]: This should be GitHub Packages right?

File: .github/workflows/publish-image.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@9ec57ed1fcdbf14dcef7dfbe97b2010124a938b7
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+          labels: |
+            org.opencontainers.image.description=GOV.UK One Login Simulator
+            org.opencontainers.image.version=${{ env.RELEASE_DATE }}.${{ github.run_number }}
+          tags: ${{ env.RELEASE_DATE }}.${{ github.run_number }}
```
Comment (CarlyG55) [inline .github/workflows/publish-image.yml:47]: [from the discovery output](https://docs.google.com/document/d/1u-xKH3b19CdA4XTpeKAGM53clWtdHL0kJYmN-uE3dKc/edit#heading=h.tdwiw5dhwh9), we want to generate a git tag as well as a docker tag, and then generate the next tag from the previous one - ie not just the run number

File: .github/workflows/publish-image.yml (hunk 1 above)
Comment (gn-GDS) [reply .github/workflows/publish-image.yml:11]: Yep, this is just to trigger the actions during testing

Comment (CarlyG55) [reply .github/workflows/publish-image.yml:11]: Yeah I realised after this that you undrafted it to test, and it's probably not actually ready for review - sorry!

File: .github/workflows/publish-image.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+          push: true
+          tags: ${{ steps.meta.outputs.tags }}
+          labels: ${{ steps.meta.outputs.labels }}
+
+      - name: Generate artifact attestation
+        uses: actions/attest-build-provenance@v1
+        with:
+          subject-name: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME}}
+          subject-digest: ${{ steps.push.outputs.digest }}
+          push-to-registry: true
```
Comment (CarlyG55) [inline .github/workflows/publish-image.yml:74]: nice 🙂

File: .github/workflows/publish-image.yml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@8e5442c4ef9f78752691e2d8f8d19755c6f78e81
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+          labels: |
+            org.opencontainers.image.description=GOV.UK One Login Simulator
+            org.opencontainers.image.version=${{ env.VERSION }}
+          tags: ${{ env.VERSION }}
```
Comment (CarlyG55) [inline .github/workflows/publish-image.yml:58]: ooh could you also tag with `latest` please?

**Review by CarlyG55 — COMMENTED** [review]
just one small thing, looks good though!

**Review by CarlyG55 — APPROVED** [review]

**Review by CarlyG55 — APPROVED** [review]

**Pull request merged**